### DISCUSSION TITLE
How to know if a user has closed the webpage?

### DISCUSSION THREAD

**Discussion by Anindya088** [zauberzeug/nicegui#1379] [Q&A] [answered]
### Question

Hi All,

I am trying to create some usage stats for users, like how much time they spend on the app etc. But the challenge for me is although I can start logging the time when they start (ie. login and the session state starts), I don't know when they end, unless the specifically press the logout button.

But most often the users will just close the browser page. Can I know on the python side if a user has closed the webpage, and then do some stuffs, like log the time, or force log-off (meaning in this case, clearing the session state dictionary), etc.?

I read about some js events like ```beforeunload``` but I have no knowledge of how to send the data back to python using a js code like that. So what I found is this below:
```javascript
window.addEventListener('beforeunload', function (e) {
    e.preventDefault();
    e.returnValue = '';
});
```
Can you please help me with the correct approach here, how to know on my python side that the user is closing the webpage?

Also, related to this, how does this impact if he has multiple browser tabs open? Because ideally I would like to know, if at all even a single browser tab is open, then probably he is still working on the app. So only force log out when no browser tabs are open at all.

Thanks a lot for all the help with this.
Anindya

**Comment by rodja** [comment]
You can simply await the lifetime events of a client connection (eg. tab):

```py
from nicegui import ui, Client

@ui.page('/')
async def index(client: Client):
    print('preparing')
    await client.connected()
    print('connected')
    await client.disconnected()
    print('disconnected')

ui.run()
```

> if at all even a single browser tab is open, then probably he is still working on the app

You can use the above in combination with [`app.storage.user`](https://nicegui.io/documentation/storage) to keep track of all connect/disconnects.

**Reply by Anindya088** [reply]
Hi @rodja , 

Thanks so much, this is so much simpler than I was thinking 😄 , probably I was just over-complicating the idea because I don't know how these things work actually. 

I am trying to implement the logic that you gave together with the ```app.storage.user```, but getting one error, can you kindly advise what I am doing wrong here.

This is what I tried to create:
```python
from nicegui import ui, app, Client

@ui.page('/')
async def index(client: Client):

    print('preparing')

    # Connection start
    await client.connected()
    # Start page count in session state
    if 'PageCount' in app.storage.user:
        app.storage.user['PageCount'] = app.storage.user['PageCount'] + 1
    else:
        app.storage.user['PageCount'] = 1

    print('connected')

    # My code for the page (replace with actual later)
    name = ui.input('username')
    ui.label('Print this')

    # Connection end
    await client.disconnected()
    # Decrease page count by 1
    app.storage.user['PageCount'] = app.storage.user['PageCount'] - 1

    print('disconnected')
    print(f"Total pages = {app.storage.user['PageCount']}")

ui.run(port=5003, storage_secret='1234')
```

But this is giving some error in the console, I am not sure what is wrong here. I am sharing the error messages here below, in case you need to refer to them:
```python
NiceGUI ready to go on http://localhost:5003, http://172.18.176.1:5003, http://172.23.112.1:5003, http://172.25.144.1:5003, http://172.26.144.1:5003, http://172.27.48.1:5003, and http://192.168.18.3:5003
preparing
Expecting value: line 1 column 1 (char 0)
Traceback (most recent call last):
  File "C:\Users\anindya.chakrabort\GfK\GfK APAC MMM - Documents\Python\MMM\Lib\site-packages\nicegui\background_tasks.py", line 55, in _handle_task_result
    task.result()
  File "C:\Users\anindya.chakrabort\GfK\GfK APAC MMM - Documents\Python\MMM\Lib\site-packages\nicegui\page.py", line 91, in wait_for_result
    return await result
           ^^^^^^^^^^^^
  File "C:\Users\anindya.chakrabort\GfK\GfK APAC MMM - Documents\Python\MMM\app\modules\.test\adhoc_tests.py", line 39, in index
    if 'PageCount' in app.storage.user:
                      ^^^^^^^^^^^^^^^^
  File "C:\Users\anindya.chakrabort\GfK\GfK APAC MMM - Documents\Python\MMM\Lib\site-packages\nicegui\storage.py", line 119, in user
    self._users[id] = PersistentDict(self.storage_dir / f'storage_user_{id}.json')
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\anindya.chakrabort\GfK\GfK APAC MMM - Documents\Python\MMM\Lib\site-packages\nicegui\storage.py", line 44, in __init__
    data = json.loads(filepath.read_text()) if filepath.exists() else {}
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python\Python311\Lib\json\__init__.py", line 346, in loads
    return _default_decoder.decode(s)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python\Python311\Lib\json\decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python\Python311\Lib\json\decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

Kindly advise me on the right approach please.

Thanks a lot,
Anindya

**Reply by rodja** [reply]
Looks like the storage file is broken. Have a look into the `.nicegui` folder or try another browser / incognito mode.

**Reply by Anindya088** [reply]
Hi @rodja ,

You are right, I tested in the incognito mode for both Chrome and Edge, and both works, I can see the counts getting updated in console, as I open more tabs. But when I open the url on normal mode, both for Chrome and Edge I am getting that error.

Is there a way to fix this please? I just now updated to 1.3.8 so that I can have the latest files. But anything else I can do to fix this?

Thanks,
Anindya

**Reply by rodja** [reply, marked as answer]
>  But when I open the url on normal mode, both for Chrome and Edge I am getting that error.

You could clear the session cookie in the browser. But it would be more helpful to investigate what is broken and why. For that you should look through the storage files in the `.nicegui` directory which is automatically created in the working folder where you start your app. Do you see any non-json-conform content?

**Reply by Anindya088** [reply]
Hi @rodja ,

Thanks a lot, I cleared the browser history and then now it's working fine. So I think that was the issue.
I also checked the ```.nicegui``` folder and it looks fine now, just has the storage .json files with the ```PageCount``` variable that I am saving inside.

Also, side note, I also realized that for different browsers, the ```.nicegui``` folder will create separate .json files and manage the counts separately. So in my case, let's say my user can open different browsers, and multiple tabs in them (just thinking of the possibility, probably won't do like this), so I should save the ```username``` and ```PageCount``` both of them in the ```user.storge```, so that I can loop over the .json files and check the counts for all of them where the username matches. I hope this would be a correct approach for this.

Thanks a lot again for the help with this.
Anindya

**Reply by rodja** [reply]
> so that I can loop over the .json files and check the counts for all of them where the username matches

That does not sound right. You should not need to access the json files directly. If you want to store the counts per username I suggest you use a database or `app.storage.general`.

**Reply by Anindya088** [reply]
Hi @rodja , oh right, thanks for that, now I understand the difference between the storage variables better 😄 

I tested out the ```app.storage.general``` and yes, it's working exactly like you said, so I don't need to access the json files anymore. Thanks a lot. While working on this, I came across another question, but let me open up a separate Q/A to keep the topics separate. 

Thanks a lot for all the help.
Anindya
